src/utils/daos_hdlr.c

```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		*_cont_name = NULL;
+		rc = -ENOMEM;
+		goto out;
+	}
+	out:
+		if (f1)
+			free(f1);
+		if (f2)
+			free(f2);
+	return rc;
+}
+
+static ssize_t
+daos_write(struct daos_file_t *daos_file, char *file, void *buf, size_t size)
+{
+	int rc;
+	/* record address and size of user buffer in io vector */
+	d_iov_t iov;
+	d_iov_set(&iov, buf, size);
+
+	/* define scatter-gather list for dfs_write */
+	d_sg_list_t sgl;
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_write(daos_file->dfs, daos_file->obj, &sgl, daos_file->offset, NULL); 
+
+	if (rc) {
+		fprintf(stderr, "dfs_write %s failed (%d %s)",
+			file, rc, strerror(rc));
+		errno = rc;
+		size = -1;
+	}
+
+	daos_file->offset += (daos_off_t)size;
+	return (ssize_t) size;
+}
+
+static ssize_t
+daos_file_write(daos_file_t *daos_file, char* file,
+		void* buf, size_t size)
+{
+	ssize_t num_bytes_written = 0;
+	if (daos_file->type == POSIX) {
+		num_bytes_written = write(daos_file->fd, buf, size);
+	} else if (daos_file->type == DAOS) {
+		num_bytes_written = daos_write(daos_file, file, buf, size);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d",
+			file, daos_file->type);
+	}
+	if (num_bytes_written < 0) {
+		fprintf(stderr, "write error on %s type=%d\n",
+			file, daos_file->type);
+	}
+	return num_bytes_written;
+}
+
+static void 
+daos_open(daos_file_t *daos_file, char *file, int flags, mode_t mode)
+{
+
+	int rc; 
+	dfs_obj_t *parent = NULL;
+	char *name = NULL, *dir_name = NULL;
+
+	parse_filename(file, &name, &dir_name);
+
+	assert(dir_name);
+
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+	}
+
+	rc = dfs_open(daos_file->dfs, parent, name, mode | S_IFREG,
+		flags, 0, 0, NULL, &(daos_file->obj));
+	if (rc)
+		fprintf(stderr, "dfs_open %s failed (%d)\n", name, rc);
+}
+
+void
+daos_file_open(daos_file_t *daos_file, char* file, int flags, ...)
+{
+	/* extract the mode */
+	int mode_set = 0;
+	mode_t mode  = 0;
+	if (flags & O_CREAT) {
+		va_list ap;
+		va_start(ap, flags);
+		mode = va_arg(ap, mode_t);
+		va_end(ap);
+		mode_set = 1;
+	}
+
+	if (daos_file->type == POSIX) {
+		if (mode_set) {
+			daos_file->fd = open(file, flags, mode);
+		} else {
+			daos_file->fd = open(file, flags);
+		}
+	} else if (daos_file->type == DAOS) {
+		daos_open(daos_file, file, flags, mode);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d",
+			file, daos_file->type);
+	}
+}
+
+static int
+daos_mkdir(daos_file_t *daos_file, const char *path)
+{
+	int rc = 0;
+	dfs_obj_t *parent = NULL;
+	char *name = NULL;
+	char *dname = NULL;
+	parse_filename(path, &name, &dname);
+    
+	assert(dname);
+	if (name && strcmp(name, "/") != 0) {
+		rc = dfs_lookup(daos_file->dfs, dname, O_RDWR, &parent, NULL, NULL);
+		if (parent == NULL)
+			fprintf(stderr, "dfs_lookup %s failed \n", dname);
+		rc = dfs_mkdir(daos_file->dfs, parent, name, S_IRWXU, 0);
```

Comment:
I wonder if we need a way to pass `mode` here, instead of always using `S_IRWXU`?

Reply:
@daltonbohning yeah, not sure. I could pass in the mode from the source directory I suppose.

Reply:
Since dfs doesn't do access checks on the permissions - it just stores them - I think this could work for a dfs destination.
But in the general case, such as POSIX, we can't immediately pass the permissions from the source, because if the source doesn't have write permission, then the destination will not either.
mpifileutils handles this by updating all permissions after copying.
I'll have to give this some more thought...

Reply:
@daltonbohning I do think this is something to think about. But if the source doesn't have write permission, would we want the destination to have it?

Reply:
Right, we _**don't**_ want the destination to have write permission if the source doesn't have it.
But for POSIX, we have to set write permission on the destination in order to write to the file. Then, _**after**_ copying, we can set the destination permissions to match the source exactly.
Of course, for a destination DAOS container we can just set the permissions immediately since DFS doesn't check the access - it just stores the POSIX permissions.

So, in a general case, we can't call, say, `daos_file_mkdir` with the source permissions. This works for DAOS but not for POSIX.

Reply:
@daltonbohning ah yeah I see what you mean. We can do something similar.

Reply:
@dsikich If we aren't yet going to handle this as discussed, would you please leave a TODO comment or something similar? There aren't yet datamover tests for verifying that permissions are correct, so I'm worried this might slip through accidentally.

Reply:
@daltonbohning yes, definitely. I added one.

Reply:
Thanks!

Reply:
well it should be the same as the file create function, so we have to add a mode parameter from the calling function